pkg/util/strategicpatch/patch.go

```diff
@@ -31,7 +31,352 @@ import (
+			if err != nil {
+				return nil, err
+			}
+
+			if len(pv) > 0 {
+				patch[key] = pv
+			}
+		case string, float64, bool:
+			if !matchesValue(av, bv) {
+				patch[key] = bv
+			}
+		case []interface{}:
+			bt := bv.([]interface{})
+			if fieldPatchStrategy == "merge" {
+				pv, err := diffLists(at, bt, fieldType.Elem(), fieldPatchMergeKey)
+				if err != nil {
+					return nil, err
+				}
+
+				if len(pv) > 0 {
+					patch[key] = pv
+				}
+			} else if !matchesArray(at, bt, t) {
+				patch[key] = bv
+			}
+		case nil:
+			switch bv.(type) {
+			case nil:
+				// Both nil, fine.
+			default:
+				patch[key] = bv
+			}
+		default:
+			panic(fmt.Sprintf("Unknown type:%T in key %s", av, key))
+		}
+	}
+
+	// Now add all deleted values as nil
+	for key := range a {
+		_, found := b[key]
+		if !found {
+			patch[key] = nil
+		}
+	}
+
+	return patch, nil
+}
+
+// Returns true if the array matches (must be json types).
+// As is idiomatic for go, an empty array is not the same as a nil array.
+func matchesArray(a, b []interface{}, t reflect.Type) bool {
+	if len(a) != len(b) {
+		return false
+	}
+
+	if (a == nil && b != nil) || (a != nil && b == nil) {
+		return false
+	}
+
+	for i := range a {
+		if !matchesValue(a[i], b[i]) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// Returns true if the values matches (must be json types)
+// The types of the values must match, otherwise it will always return false
+// If two map[string]interface{} are given, all elements must match.
+func matchesValue(av, bv interface{}) bool {
+	switch at := av.(type) {
+	case string:
+		bt, ok := bv.(string)
+		if ok && (bt == at) {
+			return true
+		}
+	case float64:
+		bt, ok := bv.(float64)
+		if ok && (bt == at) {
+			return true
+		}
+	case bool:
+		bt, ok := bv.(bool)
+		if ok && (bt == at) {
+			return true
+		}
+	case map[string]interface{}:
+		bt, ok := bv.(map[string]interface{})
+		if ok && (len(at) == len(bt)) {
+			for key := range at {
+				if !matchesValue(at[key], bt[key]) {
+					return false
+				}
+			}
+
+			return true
+		}
+	}
+
+	return false
+}
+
+// Returns a (recursive) strategic merge patch that yields b when applied to a,
+// for a pair of lists with merge semantics.
+func diffLists(a, b []interface{}, t reflect.Type, mergeKey string) ([]interface{}, error) {
+	if len(a) == 0 {
+		if len(b) == 0 {
+			return nil, nil
+		}
+
+		return b, nil
+	}
+
+	et, err := sliceElementType(a, b)
+	if err != nil {
+		return nil, err
+	}
+
+	var patch []interface{}
+
+	// If the elements are not maps...
+	if et.Kind() == reflect.Map {
+		patch, err = diffListsOfMaps(a, b, t, mergeKey)
+	} else {
+		patch, err = diffListsOfScalars(a, b)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return patch, nil
+}
+
+// Returns a (recursive) strategic merge patch that yields b when applied to a,
+// for a pair of lists of scalars with merge semantics.
+func diffListsOfScalars(a, b []interface{}) ([]interface{}, error) {
```

Comment:
There are some 2 cases here where the return seems wrong but I don't know what the expected result should be:
1.
b is a subset of a e.g.
```
a = [1, 2, 3, 4]
b = [1, 2, 3]
```
in this case the diff would be an empty list, which seems wrong. But then, I don't know how you are supposed to represent removing elements from a list as a diff. What is the expected return value here?
2.
b differs from a only in number of duplicates, e.g.
```
a = [1, 2, 3]
b = [1, 2, 2, 2, 3]
```
If I'm reading this correctly, the code below will again produce an empty list for the diff because `uniqifyAndSortScalars` strips duplicates. Is there a guarantee that lists of scalars will not contain duplicates?